File: balboa-admin/src/main/java/com/socrata/balboa/admin/tools/Lister.java
package com.socrata.balboa.admin.tools;

import com.socrata.balboa.metrics.data.CompoundIterator;
import com.socrata.balboa.metrics.data.DataStore;
import com.socrata.balboa.metrics.data.DataStoreFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * List all the entity keys
 */
public class Lister {

    public void list(List<String> filters) throws IOException {
        {
            DataStore ds = DataStoreFactory.get();
            Iterator<String> entities;
            if (filters.size() > 0)
            {
                List<Iterator<String>> iters = new ArrayList<>(filters.size());
                for (String filter : filters)
                {
                    iters.add(ds.entities(filter));
                }
                entities = new CompoundIterator<>(iters);
            }
            else
            {
                entities = ds.entities();
            }
            while (entities.hasNext())
            {
                System.out.println(entities.next());
            }
        }
    }
}

File: balboa-admin/src/main/scala/com/socrata/balboa/admin/tools/Lister.scala
package com.socrata.balboa.admin.tools

import java.io.IOException

import com.socrata.balboa.metrics.data.DataStoreFactory
import scala.collection.JavaConverters._

/**
  * List all the entity keys
  */
class Lister(dataStoreFactory: DataStoreFactory) {

  def listJava(filters: java.util.Iterator[String]): Unit = list(filters.asScala)

  @throws[IOException]
  def list(filters: Iterator[String], printer: (String => Unit) = println): Unit = { // scalastyle:ignore
    val ds = dataStoreFactory.get
    val entities = (if (filters.isEmpty) {
      ds.entities()
    } else {
      filters.flatMap(filter => ds.entities(filter))
    }).toList.distinct

    entities.foreach(printer)
  }
}
